core/templates/services/audio-player.service.ts

```diff
@@ -16,162 +16,219 @@
+      this._currentTrack.seek(this._lastPauseOrSeekPos);
+    }
+    this.ngZone.runOutsideAngular(() => {
+      interval(500).pipe(takeUntil(
+        this._stopIntervalSubject)).subscribe(() => {
+        this.ngZone.run(() => {
+          this._updateViewEventEmitter.emit();
+        });
+      });
+    });
+    this._currentTrack.play();
+  }
+
+  pause(): void {
+    if (!this.isPlaying()) {
+      return;
+    }
+    this._lastPauseOrSeekPos = this.getCurrentTime();
+    this._currentTrack.pause();
+    this._stopIntervalSubject.next();
+  }
+
+  stop(): void {
+    if (!this._currentTrack) {
+      return;
+    }
+    this._lastPauseOrSeekPos = 0;
+    this._currentTrack.stop();
+    this._stopIntervalSubject.next();
+    this._currentTrackFilename = null;
+    this._currentTrack = null;
+  }
+
+  rewind(seconds: number): void {
+    if (!this._currentTrack) {
+      return;
+    }
+    const currentSeconds = (
+      this._currentTrack.seek());
+    if (typeof currentSeconds !== 'number') {
+      return;
+    }
+    if (currentSeconds - seconds > 0) {
+      this._currentTrack.seek(currentSeconds - seconds);
+    } else {
+      this._currentTrack.seek(0);
+    }
+  }
+
+  forward(seconds: number): void {
+    if (!this._currentTrack) {
+      return;
+    }
+    const currentSeconds = this._currentTrack.seek();
+    if (typeof currentSeconds !== 'number') {
+      return;
+    }
+    if (currentSeconds + seconds < this._currentTrack.duration()) {
+      this._currentTrack.seek(currentSeconds + seconds);
+    }
+  }
+
+  getCurrentTime(): number {
+    if (!this._currentTrack) {
+      return 0;
+    }
+    const sec = this._currentTrack.seek();
+    if (typeof sec !== 'number') {
+      return 0;
+    }
+    return Math.floor(sec);
+  }
+
+  setCurrentTime(val: number): void {
+    if (!this._currentTrack) {
+      return;
+    }
+    if (val < 0) {
+      this._currentTrack.seek(0);
+      return;
+    }
+    if (val > this._currentTrack.duration()) {
+      this._currentTrack.seek(this._currentTrack.duration());
+      return;
+    }
+    this._lastPauseOrSeekPos = val;
+    this._currentTrack.seek(Math.floor(val));
+  }
+
+  getAudioDuration(): number {
+    if (this._currentTrack) {
+      return this._currentTrack.duration();
+    } else {
+      return 0;
+    }
+  }
+
+  getProgress(): number {
+    if (!this._currentTrack) {
+      return 0;
+    }
+    return (
+      this._currentTrack.seek() as number) / this._currentTrack.duration();
+  }
+
+  setProgress(progress: number): void {
+    if (progress < 0 || progress > 1) {
+      return;
+    }
+    if (this._currentTrack) {
+      this._currentTrack.seek(progress * this._currentTrack.duration());
+    }
+  }
+
+  isPlaying(): boolean {
+    return this._currentTrack && this._currentTrack.playing();
+  }
+
+  isTrackLoaded(): boolean {
+    return this._currentTrack !== null;
+  }
+
+  clear(): void {
+    if (this._currentTrack && this.isPlaying()) {
```

Comment:
`this.isPlaying` is internally doing the `this._currentTrack` check, right?

Reply:
Yes... changed it.